### PR TITLE
Add `main` decorator for simple parsing of entry functions

### PR DESCRIPTION
**Pull request by JesseFarebro**
Adds a decorator `main` that will parse command line arguments based on the type annotations of that function.

This decorator supports:
- Positional arguments via Python 3.8+'s support for positional-only arguments with the `/` syntax.
- Nested dataclasses as arguments to a function.
- Automatically parsing the help string based on the docstring of the function.

Example:

```py
import pathlib

import simple_parsing as sp

@sp.decorators.main
def mkdir(path: pathlib.Path, /, *, parents: bool = False, exist_ok: bool = False) -> None:
    """ Make a directory.
    
    Args:
        path: Path at which to create the directory.
        parents: Whether to create parent directories.
        exist_ok: If exist_ok is True `FileExistsError` will be raised.
    """
    path.mkdir(parents=parents, exist_ok=exist_ok)
    
if __name__ == '__main__':
    mkdir()
```

### CODE REVIEW

File: simple_parsing/decorators.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                default, default_factory = dataclasses.MISSING, dataclasses.MISSING
+                if parameter.default != inspect.Parameter.empty:
+                    if inspect.isfunction(parameter.default):
+                        default_factory = parameter.default
+                    else:
+                        default = parameter.default
+
+                field = _Field(
+                    name,
+                    parameter.annotation,
```
Comment (lebrice) [inline simple_parsing/decorators.py:66]: What happens if the user script is using postponed evaluation of type annotations (`from __future__ import annotations`)? I assume that `parameter.annotation` would then be a string. Would this code be affected by that?

The rest of simple-parsing can deal fairly well with postponed evaluations, by evaluating them back into a runtime type annotations (e.g. `"int | bool"` gets brought back down to a `typing.Union[int, bool]` later down the line, you don't have to worry about doing that here)

File: test/test_decorator.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        ("", 3, partial(_fn_with_all_argument_types, 1, b=1, c=1)),
+        ("2", 4, partial(_fn_with_all_argument_types, b=1, c=1)),
+        ("2 --b=2", 5, partial(_fn_with_all_argument_types, c=1)),
+        ("2 --b=2 --c=2", 6, _fn_with_all_argument_types),
+        ("--c=2", 4, partial(_fn_with_all_argument_types, 1, b=1)),
+        ("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
+        ("--b=2 --c=2", 5, partial(_fn_with_all_argument_types, 1)),
+    ],
+)
+def test_simple_arguments(
```
Comment (lebrice) [inline test/test_decorator.py:88]: These tests are fine for now, but in the future we'd need to make sure this works with other types of fields besides ints.

Just for fun, here's how I parameterize the field types for other tests, which I think is both very hacky and very pretty:

https://github.com/lebrice/SimpleParsing/blob/master/test/test_base.py#L81

Perhaps you could do something similar when constructing your functions (or I could add it myself after this PR).
Note: If I recall correctly, this kind of trickery only works when you type annotations are evaluated at runtime (as you are doing here).

**Review by lebrice — APPROVED** [review]
Thanks! Left a few suggestions, but feel free to ignore them if you want.

File: simple_parsing/__init__.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 """Simple, Elegant Argument parsing.
 @author: Fabrice Normandin
 """
-from . import helpers, utils, wrappers
+from . import decorators, helpers, utils, wrappers
```
[the remark this answers is not in the document]
Comment (JesseFarebro) [reply simple_parsing/__init__.py:4]: Yeah, I'd prefer this as well.

File: simple_parsing/decorators.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+class _Field(NamedTuple):
+    name: str
+    annotation: Type
+    field: dataclasses.Field
+
+
+def main(original_function: Optional[Callable] = None, **sp_kwargs) -> Callable:
+    """Parse a function's arguments using simple-parsing from type annotations."""
+
+    def _description_from_docstring(docstring: dp.Docstring) -> str:
```
[the remark this answers is not in the document]
Comment (JesseFarebro) [reply simple_parsing/decorators.py:21]: Good catch!

File: simple_parsing/decorators.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+from . import helpers, parsing
+
+
+class _Field(NamedTuple):
+    name: str
+    annotation: Type
+    field: dataclasses.Field
+
+
+def main(original_function: Optional[Callable] = None, **sp_kwargs) -> Callable:
```
[the remark this answers is not in the document]
Comment (JesseFarebro) [reply simple_parsing/decorators.py:18]: Yeah, exactly. You can decorate in two ways:

```
@sp.main

@sp.main(**kwargs)
```
where `**kwargs` will be forwarded to `sp.parse`.

File: test/test_decorator.py (hunk 2 above)
Comment (JesseFarebro) [reply test/test_decorator.py:88]: The reason I hadn't tested other fields is that those tests should already be covered by SP. We need to check that the decorator constructs the dataclass properly and everything else is handled by `sp.parse`. If we have test coverage for those types in `sp.parse` we should be good.

Comment (lebrice) [reply test/test_decorator.py:88]: Fair enough.

File: simple_parsing/decorators.py (hunk 1 above)
Comment (lebrice) [reply simple_parsing/decorators.py:77]: Re: We can take a look at postponed evaluations later. This should be fine for now.

**Pull request merged**